sklearn/decomposition/sparse_pca.py

```diff
@@ -13,6 +13,61 @@
 from .dict_learning import dict_learning, dict_learning_online
 
 
+def get_explained_variance(X, components):
+    '''
+        Get the explained variance.
+
+        Get the explained variance from the principal components of the
+        data.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The feature vector. n_samples and n_features are the number of
+            samples and features, respectively.
+
+        components : array, shape (n_components, n_features)
+            The (un-normalized) principle components. [1]
+
+        Notes
+        -----
+        The variance ratio may not be computed. The main reason is that we
+        do not know what the total variance is since we did not compute all
+        the components.
+
+        References
+        ----------
+        [1] https://web.stanford.edu/~hastie/Papers/spc_jcgs.pdf
+    '''
+    # the number of samples
+    n_samples = X.shape[0]
+    n_components = components.shape[0]
+    unit_vecs = components.copy()
+    components_norm = np.linalg.norm(components, axis=1)[:, np.newaxis]
+    components_norm[components_norm == 0] = 1
+    unit_vecs /= components_norm
+
+    # Algorithm, as we compute the variance for each projection, we
+    # subtract the variance from components in the direction of previous
+    # axes
+    proj_corrected_vecs = np.zeros_like(components)
+    for i in range(n_components):
+        vec = components[i].copy()
+        # subtract the previous projections
+        for j in range(i):
+            vec -= np.dot(unit_vecs[j], vec)*unit_vecs[j]
+
+        proj_corrected_vecs[i] = vec
+
+    # get estimated variance of Y which is matrix product of feature vector
+    # and the components (projection corrected)
```

Comment:
I am not sure "projection corrected" is the right term here

Reply:
changed to adjusted